rust/otap-dataflow/crates/otap/src/otap_grpc.rs

```diff
@@ -224,162 +168,186 @@ impl ArrowTracesService for ArrowTracesServiceImpl {
+    input_stream: tonic::Streaming<BatchArrowRecords>,
+    effect_handler: shared::EffectHandler<OtapPdata>,
+    state: Option<AckSubscriptionState>,
     otap_batch: F,
-    consumer: &mut Consumer,
-    mut batch: BatchArrowRecords,
-    effect_handler: &shared::EffectHandler<OtapPdata>,
-    state: Option<SharedState>,
-    tx: &tokio::sync::mpsc::Sender<Result<BatchStatus, Status>>,
-) -> Result<(), ()>
+) -> Pin<Box<dyn Stream<Item = Result<BatchStatus, Status>> + Send + 'static>>
 where
-    F: Fn(T) -> OtapArrowRecords,
+    T: OtapBatchStore + 'static,
+    F: Fn(T) -> OtapArrowRecords + Send + Copy + 'static,
 {
-    let batch_id = batch.batch_id;
-    let batch = consumer.consume_bar(&mut batch).map_err(|e| {
-        log::error!("Error decoding OTAP Batch: {e:?}. Closing stream");
-    })?;
-
-    let batch = from_record_messages::<T>(batch);
-    let otap_batch_as_otap_arrow_records = otap_batch(batch);
-    let mut otap_pdata =
-        OtapPdata::new(Context::default(), otap_batch_as_otap_arrow_records.into());
-
-    let cancel_rx = if let Some(state) = state {
-        // Try to allocate a slot (under the mutex) for calldata.
-        let allocation_result = {
-            let guard_result = state.0.lock();
-            match guard_result {
-                Ok(mut guard) => guard.allocate(|| oneshot::channel()),
-                Err(_) => {
-                    log::error!("Mutex poisoned");
-                    return Err(());
+    struct ArrowBatchStreamState<T, F>
+    where
+        T: OtapBatchStore + 'static,
+        F: Fn(T) -> OtapArrowRecords + Send + Copy + 'static,
+    {
+        input_stream: tonic::Streaming<BatchArrowRecords>,
+        consumer: Consumer,
+        effect_handler: shared::EffectHandler<OtapPdata>,
+        state: Option<AckSubscriptionState>,
+        otap_batch: F,
+        finished: bool,
+        _marker: PhantomData<fn() -> T>,
+    }
+
+    enum StreamStep {
+        Yield(Result<BatchStatus, Status>),
+        Done,
+    }
+
+    enum BatchProcessingResult {
+        Emit(BatchStatus),
+        Terminate,
+    }
+
+    impl<T, F> ArrowBatchStreamState<T, F>
+    where
+        T: OtapBatchStore + 'static,
+        F: Fn(T) -> OtapArrowRecords + Send + Copy + 'static,
+    {
+        async fn next_item(&mut self) -> StreamStep {
+            if self.finished {
+                return StreamStep::Done;
+            }
+
+            match self.input_stream.message().await {
+                Ok(Some(batch)) => match self.process_batch(batch).await {
+                    BatchProcessingResult::Emit(status) => StreamStep::Yield(Ok(status)),
+                    BatchProcessingResult::Terminate => StreamStep::Done,
+                },
+                Ok(None) => {
+                    self.finished = true;
+                    StreamStep::Done
+                }
+                Err(status) => {
+                    self.finished = true;
+                    StreamStep::Yield(Err(status))
```

Comment:
Do we always want to terminate the stream on any error? Not something for this PR, but maybe we should inspect the error to check if it's a fatal error or not?

cc @jmacd